src/Kerbalism/SteppedSim/Jobs/FluxAnalysisJobs.cs

```diff
@@ -0,0 +1,380 @@
+					stars.Add(i);
+		}
+	}
+
+	[BurstCompile]
+	public struct BodySunOcclusionJob : IJobParallelFor
+	{
+		[ReadOnly] public NativeArray<int3> triplets;
+		[ReadOnly] public int bodiesPerStep;
+		[ReadOnly] public NativeArray<SubstepBody> bodies;
+		[ReadOnly] public NativeArray<int> starIndexes;
+		[WriteOnly] public NativeArray<bool> occluded;
+
+		public void Execute(int index)
+		{
+			var body = bodies[index];
+			if (Unity.Burst.CompilerServices.Hint.Unlikely(body.solarLuminosity > 0))
+				occluded[index] = false;
+			else
+			{
+				int step = triplets[index].x;
+				int firstBodyIndex = step * bodiesPerStep;
+				var frameSlice = bodies.Slice(firstBodyIndex, bodiesPerStep);
+				bool occludedTemp = true;
+				for (int i = 0; i < starIndexes.Length && !occludedTemp; i++)
+				{
+					var star = bodies[firstBodyIndex + starIndexes[i]];
+					// avoid self-occluding, so adjust the body and sun positions...
+					var dir = math.normalize(star.position - body.position);
+					var start = body.position + (body.radius + 10) * dir;
+					var end = star.position - (star.radius + 10) * dir;
+					occludedTemp &= SubstepBody.Occluded(start, end, frameSlice);
+				}
+				occluded[index] = occludedTemp;
+			}
+		}
+	}
+
+	[BurstCompile]
+	public struct BodyVesselOcclusionJob : IJobParallelFor
+	{
+		[ReadOnly] public NativeArray<SubstepVessel> vessels;
+		[ReadOnly] public NativeArray<SubstepBody> bodies;
+		[ReadOnly] public int numBodiesPerStep;
+		[ReadOnly] public NativeArray<int3> triplets;
+		[DeallocateOnJobCompletion] [ReadOnly] public NativeArray<bool> occlusionRelevance;
+		[WriteOnly] public NativeArray<bool> occluded;
+
+		public void Execute(int index)
+		{
+			var triplet = triplets[index];
+			var step = triplet.x;
+			var vessel = vessels[triplet.y];
+			var body = bodies[triplet.z];
+			// avoid self-occluding, so adjust the body position
+			var dir = math.normalize(body.position - vessel.position);
+			var end = body.position - (body.radius + 10) * dir;
+
+			var a = vessel.position;
+			var b = end;
+
+			double3 ab = b - a;
+			var abLen2 = math.lengthsq(ab);
+			if (Unity.Burst.CompilerServices.Hint.Unlikely(abLen2 < 1))
+				occluded[index] = false;
+			else
+			{
+				bool occludedLocal = false;
+				int firstBodyInStepIndex = step * numBodiesPerStep;
+				for (int testBodyIndex = firstBodyInStepIndex;
+					testBodyIndex < firstBodyInStepIndex + numBodiesPerStep;
+					testBodyIndex++)
+				{
+					if (occlusionRelevance[testBodyIndex])
+					{
+						var testBody = bodies[testBodyIndex];
+						double3 v = testBody.position;
+						var radiusSq = testBody.radius * testBody.radius;
+						double3 av = v - a;
+						double3 bv = v - b;
+						if (math.dot(av, ab) < 0)
+							occludedLocal |= math.lengthsq(av) <= radiusSq;
+						else if (math.dot(bv, ab) > 0)
+							occludedLocal |= math.lengthsq(bv) <= radiusSq;
+						else
+							occludedLocal |= math.lengthsq(math.cross(ab, av)) <= radiusSq * abLen2;
+					}
```

Comment:
I think that your occlusion check is incorrect, or presupposes a lot of things about what it is evaluating.
Your implementation requires a ray of unit length (direction) already established.
The ray must not be within the sphere when it starts.  (fix with k < -body.Radius instead of 0?)

The other code acts on an arbitrary line segment, point and distance from point.

I think these are probably similar methods, or can become the same after the assumptions in your method are backed out.  I'm more than happy if someone finds a more optimal general solution for (distance from point to line segment).

Reply:
Yes, dir must be normalized. It works perfectly well if the start point is inside the sphere radius. This code (and the various optimized variations in Sim) have worked perfectly fine for years.

Reply:
Eh?

Body --  Vessel-Facing->   ----- Body+Radius       -> Distant Target

dir is towards the right.
diff = Body.position - start is towards the left.
k = math.dot (diff, dir) is negative.
The vessel is inside the occluder.  This will answer that it is not.

Reply:
You're assuming this should return "occluded" if the point is inside the occluder, which isn't the case. That's the point of that method, it works reliably (for our purpose) at any altitude, including negative ones.